tachyon/zk/plonk/permutation/permutation_argument_runner_impl.h

```diff
@@ -111,57 +113,49 @@ PermutationArgumentRunner<Poly, Evals>::EvaluateCommitted(
 
 template <typename Poly, typename Evals>
 template <typename PCS, typename F>
-std::vector<ProverQuery<PCS>>
+std::vector<crypto::PolynomialOpening<Poly>>
 PermutationArgumentRunner<Poly, Evals>::OpenEvaluated(
-    const ProverBase<PCS>* prover, const PermutationEvaluated<Poly>& evaluated,
-    const F& x) {
+    ProverBase<PCS>* prover, const PermutationEvaluated<Poly>& evaluated,
+    const F& x, crypto::PointSet<F>& points) {
   const std::vector<BlindedPolynomial<Poly>>& product_polys =
       evaluated.product_polys();
 
-  std::vector<ProverQuery<PCS>> ret;
+  std::vector<crypto::PolynomialOpening<Poly>> ret;
   ret.reserve(product_polys.size() * 3 - 1);
 
   F x_next = Rotation::Next().RotateOmega(prover->domain(), x);
+  base::DeepRef<const F> x_next_ref = points.Insert(x_next);
+  base::DeepRef<const F> x_ref = points.Insert(x);
   for (const BlindedPolynomial<Poly>& blinded_poly : product_polys) {
-    ret.emplace_back(x, blinded_poly.ToRef());
-    ret.emplace_back(x_next, blinded_poly.ToRef());
+    const Poly& poly = blinded_poly.poly();
+    ret.emplace_back(base::DeepRef<const Poly>(&poly), x_ref, poly.Evaluate(x));
+    ret.emplace_back(base::DeepRef<const Poly>(&poly), x_next_ref,
+                     poly.Evaluate(x_next));
   }
 
-  int32_t blinding_factors =
-      static_cast<int32_t>(prover->blinder().blinding_factors());
-  F x_last = Rotation(-(blinding_factors + 1)).RotateOmega(prover->domain(), x);
+  base::CheckedNumeric<int32_t> blinding_factors =
+      prover->blinder().blinding_factors();
+  Rotation last_rotation = Rotation((-blinding_factors - 1).ValueOrDie());
```

Comment:
This change should be separated and also shouldn't be changed in this way. There's other cases that use this pattern as well.